### PR TITLE
refactor: using PAC instead of atomic units for external input/outputs

### PR DESCRIPTION
**Pull request by amirvalhalla**
## Description

convert coin type for Proto requests to accept coin instead of coin unit

### CODE REVIEW

File: util/utils.go [hunk 1]
```diff
@@ -124,18 +124,18 @@ func IS2OP(x *big.Int, xLen int) []byte {
...
+// Example: ConvertCoinToCoinUnit(2.75) returns 2750000000.
+func ConvertCoinToCoinUnit(coin float64) int64 {
+	return int64(coin * coinUnit)
 }
 
-// ChangeToCoin converts a change value to its corresponding coin amount.
-// Example: ChangeToCoin(2750000000) returns 2.75.
-func ChangeToCoin(change int64) float64 {
-	return float64(change) / changeFactor
+// ConvertCoinUnitToCoin converts a change value to its corresponding coin amount.
```
Comment (b00f) [inline util/utils.go:135]: Please don't change this function name. Now we have many changes because of this rename.

File: util/utils.go [hunk 2]
```diff
@@ -113,7 +113,7 @@ func OS2IP(os []byte) *big.Int {
 	return new(big.Int).SetBytes(os)
 }
 
-// I2OSP converts a nonnegative integer to an octet string of a specified length.
+// IS2OP converts a nonnegative integer to an octet string of a specified length.
```
Comment (b00f) [inline util/utils.go:116]: The name should be "I2OSP" ( Integer-to-Octet-String primitive)

Comment (amirvalhalla) [reply util/utils.go:116]: please check the function name. the previous name was wrong. this is the correct one.

File: wallet/client.go [hunk 3]
```diff
@@ -102,7 +103,10 @@ func (c *grpcClient) getTransaction(id tx.ID) (*pactus.GetTransactionResponse, e
 
 func (c *grpcClient) getFee(amount int64, payloadType payload.Type) (int64, error) {
 	res, err := c.transactionClient.CalculateFee(c.ctx,
-		&pactus.CalculateFeeRequest{Amount: amount, PayloadType: pactus.PayloadType(payloadType)})
+		&pactus.CalculateFeeRequest{
+			Amount:      util.ChangeToCoin(amount),
```
Comment (kehiy) [inline wallet/client.go:107]: I think we should use CoinToChainge here.

Comment (b00f) [reply wallet/client.go:107]: Do we have test for this case?

File: www/grpc/transaction_test.go [hunk 4]
```diff
@@ -124,7 +125,7 @@ func TestGetCalculateFee(t *testing.T) {
 	amount := td.RandAmount()
 	res, err := client.CalculateFee(context.Background(),
 		&pactus.CalculateFeeRequest{
-			Amount:      amount,
+			Amount:      util.ChangeToCoin(amount),
```
Comment (b00f) [inline www/grpc/transaction_test.go:128]: Revert this line

File: wallet/client.go (hunk 3 above)
Comment (kehiy) [reply wallet/client.go:107]: which case? where?

Comment (b00f) [reply wallet/client.go:107]: getFee?

Comment (kehiy) [reply wallet/client.go:107]: as I can remember it had tests. in the execute module and state module. I need to check one more time.

File: wallet/client.go [hunk 5]
```diff
@@ -102,10 +103,13 @@ func (c *grpcClient) getTransaction(id tx.ID) (*pactus.GetTransactionResponse, e
...
+			PayloadType: pactus.PayloadType(payloadType),
+		})
 	if err != nil {
 		return 0, err
 	}
 
-	return res.Fee, nil
+	return util.CoinToChange(res.Fee), nil
```
Comment (b00f) [inline wallet/client.go:114]: Change to Coin! Amir do you tests these APIs?
Please make sure we test them, honestly we don't have chance to fix them after release.
Did you read above thread?

File: www/grpc/transaction_test.go (hunk 4 above)
Comment (b00f) [reply www/grpc/transaction_test.go:128]: Amir it is unacceptable!
Explain here, why a refactor should change tests?

**Review by b00f — CHANGES_REQUESTED** [review]

File: wallet/client.go (hunk 5 above)
Comment (b00f) [reply wallet/client.go:114]: @amirvalhalla Apologize for the mistake. here is the client API, so it is correct.

File: www/grpc/proto/transaction.proto [hunk 6]
```diff
@@ -56,7 +56,7 @@ message GetTransactionResponse {
 // Request message for calculating transaction fee.
 message CalculateFeeRequest {
   // Transaction amount.
```
Comment (b00f) [inline www/grpc/proto/transaction.proto:58]:
```suggestion
  // Transaction amount in PAC.
```

File: www/grpc/proto/transaction.proto [hunk 7]
```diff
@@ -66,9 +66,9 @@ message CalculateFeeRequest {
 // Response message containing the calculated transaction fee.
 message CalculateFeeResponse {
   // Calculated amount.
```
Comment (b00f) [inline www/grpc/proto/transaction.proto:68]:
```suggestion
  // Calculated amount in PAC.
```

File: www/grpc/proto/transaction.proto [hunk 8]
```diff
@@ -66,9 +66,9 @@ message CalculateFeeRequest {
...
-  int64 amount = 2;
+  double amount = 2;
   // Calculated transaction fee.
```
Comment (b00f) [inline www/grpc/proto/transaction.proto:70]:
```suggestion
  // Calculated transaction fee in PAC.
```

File: www/grpc/proto/transaction.proto [hunk 9]
```diff
@@ -66,9 +66,9 @@ message CalculateFeeRequest {
 // Response message containing the calculated transaction fee.
 message CalculateFeeResponse {
   // Calculated amount.
-  int64 amount = 2;
+  double amount = 2;
```
Comment (b00f) [inline www/grpc/proto/transaction.proto:69]:
```suggestion
  double amount = 1;
```

File: www/grpc/proto/transaction.proto [hunk 10]
```diff
@@ -66,9 +66,9 @@ message CalculateFeeRequest {
...
   // Calculated transaction fee.
-  int64 fee = 1;
+  double fee = 1;
```
Comment (b00f) [inline www/grpc/proto/transaction.proto:71]:
```suggestion
  double fee = 2;
```

File: www/grpc/proto/transaction.proto [hunk 11]
```diff
@@ -94,10 +94,10 @@ message GetRawTransferTransactionRequest {
   string receiver = 3;
   // Transfer amount.
```
Comment (b00f) [inline www/grpc/proto/transaction.proto:95]:
```suggestion
  // Transfer amount in PAC.
```

File: www/grpc/proto/transaction.proto [hunk 12]
```diff
@@ -94,10 +94,10 @@ message GetRawTransferTransactionRequest {
...
   // It should be greater than 0.
-  int64 amount = 4;
+  double amount = 4;
   // Transaction fee.
```
Comment (b00f) [inline www/grpc/proto/transaction.proto:98]:
```suggestion
  // Transaction fee in PAC.
```

File: www/grpc/proto/transaction.proto [hunk 13]
```diff
@@ -118,7 +118,7 @@ message GetRawBondTransactionRequest {
   string public_key = 5;
   // Transaction fee.
```
Comment (b00f) [inline www/grpc/proto/transaction.proto:119]:
```suggestion
  // Transaction fee in PAC.
```

File: www/grpc/proto/transaction.proto [hunk 14]
```diff
@@ -145,10 +145,10 @@ message GetRawWithdrawTransactionRequest {
   string account_address = 3;
   // Withdrawal amount.
```
Comment (b00f) [inline www/grpc/proto/transaction.proto:146]:
```suggestion
  // Withdrawal amount in PAC.
```

File: www/grpc/proto/transaction.proto [hunk 15]
```diff
@@ -145,10 +145,10 @@ message GetRawWithdrawTransactionRequest {
...
   // It should be greater than 0.
-  int64 amount = 4;
+  double amount = 4;
   // Transaction fee.
```
Comment (b00f) [inline www/grpc/proto/transaction.proto:149]:
```suggestion
  // Transaction fee in PAC.
```

**Review by b00f — APPROVED** [review]

**Pull request merged**